crates/oxc_linter/src/rules/typescript/prefer_for_of.rs

```diff
@@ -0,0 +1,296 @@
+        | AstKind::ArrayPattern(_) => true,
+        AstKind::UnaryExpression(unary_expr) => unary_expr.operator == UnaryOperator::Delete,
+        _ => false,
+    }
+}
+
+impl Rule for PreferForOf {
+    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
+        let AstKind::ForStatement(for_stmt) = node.kind() else { return };
+
+        let Some(ForStatementInit::VariableDeclaration(for_stmt_init)) = &for_stmt.init else {
+            return;
+        };
+
+        if for_stmt_init.declarations.len() != 1
+            || for_stmt_init.kind == VariableDeclarationKind::Const
+        {
+            return;
+        }
+
+        let decl = &for_stmt_init.declarations[0];
+        let (var_name, var_symbol_id) = match &decl.id.kind {
+            BindingPatternKind::BindingIdentifier(id) => (&id.name, id.symbol_id.get()),
+            _ => return,
+        };
+
+        if !matches!(&decl.init,
+            Some(Expression::NumericLiteral(literal)) if literal.value == 0f64
+        ) {
+            return;
+        }
+
+        let Some(Expression::BinaryExpression(test_expr)) = &for_stmt.test else { return };
+
+        if !matches!((&test_expr.left, test_expr.operator),
+            (Expression::Identifier(id), BinaryOperator::LessThan) if id.name == var_name
+        ) {
+            return;
+        }
+
+        let array_name = {
+            let Expression::MemberExpression(mem_expr) = &test_expr.right else { return };
+            if !matches!(mem_expr.static_property_name(), Some(prop_name) if prop_name == "length")
+            {
+                return;
+            }
+
+            match &mem_expr.object() {
+                Expression::Identifier(id) => id.name.as_str(),
+                Expression::MemberExpression(m_expr) => match m_expr.static_property_name() {
+                    Some(array_name) => array_name,
+                    None => return,
+                },
+                _ => return,
+            }
+        };
+
+        let Some(update_expr) = &for_stmt.update else { return };
+        if !update_expr.is_increment_of(var_name) {
+            return;
+        }
+
+        let nodes = ctx.nodes();
+        let test_and_update_span = test_expr.span.merge(&update_expr.span());
+        let body_span = for_stmt.body.span();
+
+        let Some(var_symbol_id) = var_symbol_id else { return };
+        if ctx.semantic().symbol_references(var_symbol_id).any(|reference| {
+            let ref_id = reference.node_id();
+
+            let symbol_span = nodes.get_node(ref_id).kind().span();
+            if test_and_update_span.contains(symbol_span) || !body_span.contains(symbol_span) {
+                return false;
+            }
+
+            let Some(ref_parent) = nodes.parent_node(ref_id) else { return true };
+            if matches!(nodes.parent_node(ref_parent.id()), Some(ref_grand_parent)
+                if is_assignee(ref_grand_parent))
+            {
+                return true;
+            }
```

Comment:
Looks like we didn't do the same with the original [`isAssignee`](https://github.com/typescript-eslint/typescript-eslint/blob/04e32d6834c19160fb4f44f812a3e64671fca893/packages/eslint-plugin/src/util/isAssignee.ts#L13). (e.g. we didn't check it's on left side of `AssignmentExpression`). Do we need to keep the same with it?

Reply:
Checked all the cases, and fortunately for us, it looks like the following:

| Code | Grand Parent Kind |
| - | - |
| `for (let i = 0; i < arr.length; i++) { delete arr[i]; }` | UnaryExpression |
| `for (let i = 0; i < arr.length; i++) { arr[i] = 1; }` | SimpleAssignmentTarget |
| `for (let i = 0; i < arr.length; i++) { arr[i] += 1; }` | SimpleAssignmentTarget |
| `for (let i = 0; i < arr.length; i++) { arr[i]++; }` | SimpleAssignmentTarget |
| `for (let i = 0; i < arr.length; i++) { --arr[i]; }` | SimpleAssignmentTarget |
| `for (let i = 0; i < arr.length; i++) { [arr[i]] = [0]; }` | SimpleAssignmentTarget |
| `for (let i = 0; i < arr.length; i++) { [...arr[i]] = [0]; }` | SimpleAssignmentTarget |
| `for (let i = 0; i < arr.length; i++) { ({ foo: arr[i] } = { foo: 0 }) }` | SimpleAssignmentTarget |

When we assign or update the value within the array, the grandparent of `i` is identified as a `SimpleAssignmentTarget`.

As an example, in the most common form:
- `arr[i] = 1;` is an `AssignmentExpression`.
- `arr[i]` in `assign_expr.left` is categorized under the `SimpleAssignmentTarget` kind, containing a `ComputedMemberExpression` with `object` being `"arr"` and `expression` being `"i"`.

Therefore, we do not need to check the `left` side since we are already positioned on that node (on the `left`). All that is required is to verify its type.

I'm gonna to update the code and completely remove the `is_assignee` function.

Reply:
Nice approach!